kernel/aster-nix/src/process/sync/condvar.rs

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::LinkedList, sync::Arc};
+use core::{
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
+
+use aster_frame::sync::{guard_lock, MutexGuard, SpinLock};
+
+use super::Pauser;
+use crate::Errno;
+
+/// The lock error type for situations where the condvar is not notified.
+pub enum LockErr<Guard> {
+    Timeout(Guard),
+    Interrupted(Guard),
+    Unknown(Guard),
+}
+/// LockResult, different from Rust std.
+/// The result of a lock operation.
+pub type LockResult<Guard> = Result<Guard, LockErr<Guard>>;
+
+impl<Guard> LockErr<Guard> {
+    pub fn unwrap(self) -> Guard {
+        match self {
+            LockErr::Timeout(guard) => guard,
+            LockErr::Interrupted(guard) => guard,
+            LockErr::Unknown(guard) => guard,
+        }
+    }
+}
+
+/// A condition variable.
+pub struct Condvar {
+    pauser: Arc<Pauser>,
+    should_wait: SpinLock<LinkedList<Arc<AtomicBool>>>,
+}
+
+impl Condvar {
+    /// Creates a new condition variable.
+    pub fn new() -> Self {
+        Condvar {
+            pauser: Pauser::new(),
+            should_wait: SpinLock::new(LinkedList::new()),
+        }
+    }
+
+    /// Waits for the condition variable to be signaled or broadcasted.
+    pub fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> LockResult<MutexGuard<'a, T>> {
+        let lock = guard_lock(&guard);
+        drop(guard);
+        self.should_wait
+            .lock()
+            .push_back(Arc::new(AtomicBool::new(true)));
+        let cond = || {
+            let wait_list = self.should_wait.lock();
+            if let Some(wait) = wait_list.front() {
+                if wait.load(Ordering::SeqCst) {
+                    return None;
+                }
+            }
+
+            Some(())
```

Comment:
The `cond` will cause race condition.

Suppose there are two threads waiting on the same `CondVar`, each thread has a `AtomicBool` in `should_wait`. The first `AtomicBool` is for the first thread, th second `AtomicBool` is for the second thread. 
But if the `cond` of the second thread is called at first,  it will use the first `AtomicBool`! This can happen, for example, if multiple threads are notified simutaniously, but the second thread is scheduled to run at first.

So I suggest you replace the cond as
```rust
let waiter = Arc::new(AtomicBool::new(true);
self.should_wait.lock().push(waiter.clone());

let cond = | | {
    if wait.load(Ordering::SeqCst) {
          None
     } else {
          Some(())
     }
}
```

Reply:
Another related but separate concern of mine is that is it necessary to maintain a list of boolean values? Is it possible to simply use a counter? Or a per-stack local variable?

Reply:
I think there should be a simplified implementation with counters. 

`waiter_counter`: Maintains the current number of waiters.
`notify_counter`: keeps track of the number of waiters to be woken up.
When the `cond()` function executes, it checks if `notify_counter` is greater than zero. If it is, it determines that the wakeup condition has been met and allows the wait to end. Subsequently, `notify_counter` is decremented once to indicate that a waiter has been woken up.